pkg/envoy/server.go

```diff
@@ -395,7 +412,43 @@ func (s *XDSServer) AddListener(name string, kind policy.L7ParserType, port uint
 		}
 		listenerConf.FilterChains = append(listenerConf.FilterChains, s.getHttpFilterChainProto(tlsClusterName, true))
 	} else {
-		listenerConf.FilterChains = append(listenerConf.FilterChains, s.getTcpFilterChainProto(clusterName))
+		tcpChain := s.getTcpFilterChainProto(clusterName, true)
+		tcpChain.FilterChainMatch = &envoy_config_listener.FilterChainMatch{
+			// must have transport match, otherwise TLS inspector will be automatically inserted
+			TransportProtocol: "raw_buffer",
+		}
+		listenerConf.FilterChains = append(listenerConf.FilterChains, tcpChain)
+
+		mySqlChain := s.getTcpFilterChainProto(clusterName, false)
+		mySqlChain.Filters = append([]*envoy_config_listener.Filter{{
+			Name: "envoy.filters.network.mysql_proxy",
+			ConfigType: &envoy_config_listener.Filter_TypedConfig{
+				TypedConfig: toAny(&envoy_mysql_proxy.MySQLProxy{
+					StatPrefix: "mysql",
+				}),
+			}}}, mySqlChain.Filters...)
+		mySqlChain.FilterChainMatch = &envoy_config_listener.FilterChainMatch{
+			// must have transport match, otherwise TLS inspector will be automatically inserted
+			TransportProtocol:    "raw_buffer",
+			ApplicationProtocols: []string{"envoy.filters.network.mysql_proxy"},
+		}
+		listenerConf.FilterChains = append(listenerConf.FilterChains, mySqlChain)
+
+		mongoChain := s.getTcpFilterChainProto(clusterName, false)
+		mongoChain.Filters = append([]*envoy_config_listener.Filter{{
+			Name: "envoy.mongo_proxy",
+			ConfigType: &envoy_config_listener.Filter_TypedConfig{
+				TypedConfig: toAny(&envoy_mongo_proxy.MongoProxy{
+					StatPrefix:          "mongo",
+					EmitDynamicMetadata: true,
+				}),
+			}}}, mongoChain.Filters...)
+		mongoChain.FilterChainMatch = &envoy_config_listener.FilterChainMatch{
+			// must have transport match, otherwise TLS inspector will be automatically inserted
+			TransportProtocol:    "raw_buffer",
+			ApplicationProtocols: []string{"envoy.mongo_proxy"},
+		}
+		listenerConf.FilterChains = append(listenerConf.FilterChains, mongoChain)
```

Comment:
I wonder if we can extract this common code to a helper function, and pass protocol specific fields to it. That way the code won't be further duplicated as we add more protocols in future.
Something like this - 

```
func prependFilter(FilterChain* , name, typedConfig, ...) {

}
```

Reply:
Will try this out, thanks!

Reply:
Refactored :-)